src/main/java/seedu/address/model/addresses/PublicAddressesComposition.java

```diff
@@ -144,22 +144,77 @@ public PublicAddressesComposition filterByPublicAddress(String publicAddressStri
 
     /**
      * Adds a new public address to the composition.
+     * Addresses with existing labels will be replaced.
      *
      * @param newPublicAddress The new public address to be added.
      * @return A new PublicAddressesComposition with the updated public addresses.
      */
     public PublicAddressesComposition add(PublicAddress newPublicAddress) {
         assert newPublicAddress != null;
-        Map<Network, Set<PublicAddress>> updatedPublicAddresses = publicAddresses.entrySet().stream()
-            .map(entry -> {
-                Set<PublicAddress> updatedAddresses = entry.getValue().stream()
-                    .map(addr -> addr.label.equals(newPublicAddress.label)
-                        ? newPublicAddress
-                        : addr)
+
+        Map<Network, Set<PublicAddress>> updatedPublicAddresses = new HashMap<>(publicAddresses);
+
+        // Get or create the set of addresses for this network
+        Set<PublicAddress> networkAddresses = updatedPublicAddresses.getOrDefault(
+                newPublicAddress.getNetwork(),
+                new HashSet<>()
+        );
+
+        // Remove any existing address with the same label (if exists)
+        networkAddresses.removeIf(addr -> addr.label.equals(newPublicAddress.label));
+
+        // Add the new address
+        networkAddresses.add(newPublicAddress);
+
+        // Put the updated set back in the map
+        updatedPublicAddresses.put(newPublicAddress.getNetwork(), networkAddresses);
+
+        return new PublicAddressesComposition(updatedPublicAddresses);
+    }
```

Comment:
Might be better to have edit as a separate method. `add` as the name suggest should only be adding imo.